README.md
# php-simple-docmarkdowngen

Simple php markdown document generator from clases.. 
is a take of improvement for CI and Guachi first attempt .

Based and forked directly from Marco Cesarato's PHP Class Markdown Documentation

* Web Page: https://venenux.github.io/php-simple-docmarkdowngen/
* Download: https://gitlab.com/venenux/php-simple-docmarkdowngen

**YEAH!: hosted at Gitlab baby!**

## Description

Convert primitives PHPDoc comments from classes into Markdown.

By parsing a class file with a given file name 
and extracts the documentation of its functions and variables that it 
but in primitive PHPDoc format, 
then convert the extracted documentation into a file in Markdown format.

## Example

See documentation below in documentation section.

#### Usage

```php
php "include 'ClassMarkdown.php'; ClassMarkdown::printMarkdown('CoreClass.php');" >> ver.md
```


#### Output Text

```text
## CoreClass

**CoreClass, version 0.1**: Class core for all framework user base code

Copyright (c) 2018, 2019

*   PICCORO Lenz McKAY <mckaygerhard@gmail.com>

| Method        | Description                                        | Type   | Parameters                                         | Return         |
| ------------- | -------------------------------------------------- | ------ | -------------------------------------------------- | -------------- |
| __construct   | Constructor                                        | public |                                                    | void           |
| add_request   | Add Ajax request                                   |        | $name<br>$request<br>bool $public                  | bool           |
| fetch         |                                                    |        | $what<br>null $operators                           | array          |
```

# Documentaton

## ClassMarkdown

**ClassMarkdown, version 0.1.10 (2019-04-05)**: class implementation to output markdown text from class

Copyright (c) 2018, 2019

*   Marco Cesarato <cesarato.developer@gmail.com>
*   PICCORO Lenz McKAY <mckaygerhard@gmail.com>

| Method        | Description                                       | Type                  | Parameters | Return |
| ------------- | ------------------------------------------------- | --------------------- | ---------- | ------ |
| __construct   | ClassMarkdown constructor                         | public                | none       | void   |
| parseClass    | a given class funtions as array keys              | protected <br> static | $class     | array  |
| parseExtended | a given class tags as markdown formated paragraph | protected <br> static | $class     | string |
| getMarkdown   | Get markdown class documentation                  | public <br> static    | $file      | string |
| printMarkdown | Print Markdown class documentation                | public <br> static    | $file      | void   |
| getArray      | Get PHP file as array class documentation         | public <br> static    | $file      | array  |


## ClassParser

**ClassParser, **: Class ClassParser

(c) 2019


| Method                 | Description | Type   | Parameters | Return |
| ---------------------- | ----------- | ------ | ---------- | ------ |
| getClasses             |             | public |            | array  |
| getClassesImplementing |             | public | $interface | array  |
| getClassesExtending    |             | public | $class     | array  |
| parse                  |             | public | $file      | void   |


## TextTable

**TextTable, version 1.00 (2014-04-04)**: Creates a markdown table based on the parsed documentat

(c) 2019

*   Peter-Christoph Haider <peter.haider@zeyon.net>
*   PICCORO Lenz MCKAY <mckaygerhard@gmail.com>

| Method          | Description                                        | Type    | Parameters                                         | Return                                             |
| --------------- | -------------------------------------------------- | ------- | -------------------------------------------------- | -------------------------------------------------- |
| __construct     | can init the default values to markdown text table <br>  | public  | array $header  The header array [key => label, ] <br> array $content Content table as matrix of tags funtions <br> array $align   Alignment optios [key => L\|R\|C, ] | void                                               |
| setAlgin        | Overwrite the alignment array <br>                 | public  | array $align   Alignment optios [key => L\|R\|C, ] | void                                               |
| addData         | Add data to the table from matrix array tags class <br>  | public  | array $content Content matrix of tags class        | class self object class with data content populated |
| renderDelimiter | Add a delimiter based on aling of the class setted <br>  | private |                                                    | string                                             |
| renderRow       | Render a single row for markdown table <br>        | private |  array $row                                        | string                                             |
| render          | Render the table text as markdown <br>             | public  |  array  $content Additional table content          | string                                             |


Copyright (c) 2018, 2019
 
*   Marco Cesarato <cesarato.developer@gmail.com>
*   PICCORO Lenz McKAY <mckaygerhard@gmail.com>

See [LICENSE](LICENSE)

* web Page: https://venenux.github.io/php-simple-docmarkdowngen/
* Download: https://gitlab.com/venenux/php-simple-docmarkdowngen

**YEAH!: hosted at Gitlab baby!**


CHANGELOG
2019-04-04  PICCORO Lenz McKAY <mckaygerhard@gmail.com>

* very basic but support for class documentation not only functions
* only unix line feed, for new lines in generated table rows
* added funtion to exten info as paragras and funtions as lists
* make all tags not specified with default such like "void"
* limit descriptions to firts 50 chars phpdoc defines firts line as is
* property document the code
* upgrade version of class ClassMarkdown
* separate with a space the funtions tags when are new line
* property init variables if tag (such like version) does not exits
* all the title class are not level 2 "h2", was level 3
* added code documentation in the README

2018-11-22  PICCORO Lenz McKAY <mckaygerhard@gmail.com>

* initialize repo with simple README and simple target ! 
a document generator for guachi and CI powered!


TextTable.php
<?php
/**
 * Creates a markdown table based on the parsed documentation
 *
 * @author Peter-Christoph Haider <peter.haider@zeyon.net>
 * @author PICCORO Lenz MCKAY <mckaygerhard@gmail.com>
 * @package Apidoc
 * @version 1.00 (2014-04-04)
 * @license GNU Lesser Public License
 */
class TextTable {
	/** @var int max leng of descriptions */
	public $maxlen = 50;
	/** @var array data to be procesed */
	private $data = array();
	/** @var array header of the table output text */
	private $header = array();
	/** @var array The source path */
	private $len = array();
	/** @var array alingment of the headers columns */
	private $align = array(
		'name' => 'L',
		'type' => 'C'
	);

	/**
	 * can init the default values to markdown text table
	 * 
	 * @param array $header  The header array [key => label, ...]
	 * @param array $content Content table as matrix of tags funtions
	 * @param array $align   Alignment optios [key => L|R|C, ...]
	 */
	public function __construct($header=null, $content=array(), $align=false) {
		if ($header) {
			$this->header = $header;
		} elseif ($content) {
			foreach ($content[0] as $key => $value)
				$this->header[$key] = $key;
		}

		foreach ($this->header as $key => $label) {
			$this->len[$key] = strlen($label);
		}

		if (is_array($align))
			$this->setAlgin($align);

		$this->addData($content);
	}

	/**
	 * Overwrite the alignment array
	 *
	 * @param array $align   Alignment optios [key => L|R|C, ...]
	 */
	public function setAlgin($align) {
		$this->align = $align;
	}

	/**
	 * Add data to the table from matrix array tags class
	 *
	 * @param array $content Content matrix of tags class
	 * @return class self object class with data content populated
	 */
	public function addData($content) {
		foreach ($content as &$row) {
			foreach ($this->header as $key => $value) {
				if (!isset($row[$key])) {
					$row[$key] = '-';
				} elseif (strlen($row[$key]) > $this->maxlen) {
					$this->len[$key] = $this->maxlen;
				} elseif (strlen($row[$key]) > $this->len[$key]) {
					$this->len[$key] = strlen($row[$key]);
				}
			}
		}

		$this->data = $this->data + $content;
		return $this;
	}

	/**
	 * Add a delimiter based on aling of the class setted
	 *
	 * @return string
	 */
	private function renderDelimiter() {
		$res = '|';
		foreach ($this->len as $key => $l)
			$res .= (isset($this->align[$key]) && ($this->align[$key] == 'C' || $this->align[$key] == 'L') ? ':' : ' ')
			        .str_repeat('-', $l)
			        .(isset($this->align[$key]) && ($this->align[$key] == 'C' || $this->align[$key] == 'R') ? ':' : ' ')
			        .'|';
		return $res."\n";
	}

	/**
	 * Render a single row for markdown table
	 *
	 * @param  array $row
	 * @return string
	 */
	private function renderRow($row) {
		$res = '|';
		foreach ($this->len as $key => $l) {
			$res .= ' '.$row[$key].($l > strlen($row[$key]) ? str_repeat(' ', $l - strlen($row[$key])) : '').' |';
		}

		return $res."\n";
	}

	/**
	 * Render the table text as markdown
	 *
	 * @param  array  $content Additional table content
	 * @return string
	 */
	public function render($content=array()) {
		$this->addData($content);

		$res = $this->renderRow($this->header).$this->renderDelimiter();
		foreach ($this->data as $row)
			$res .= $this->renderRow($row);

		return $res;
	}
}


ClassMarkdown.php
<?php
include_once(__DIR__ . '/ClassParser.php');
include_once(__DIR__ . '/TextTable.php');

/**
 * class implementation to output markdown text from class file comments
 * 
 * @author Marco Cesarato <cesarato.developer@gmail.com>
 * @author PICCORO Lenz McKAY <mckaygerhard@gmail.com>
 * @copyright Copyright (c) 2018, 2019
 * @license http://opensource.org/licenses/gpl-3.0.html GNU Public License
 * @link https://github.com/marcocesarato/PHP-Class-Markdown-Docs
 * @version 0.1.10 (2019-04-05)
 */
class ClassMarkdown
{

    /**
     * ClassMarkdown constructor
     * @param none
     */
    public function __construct()
    {
    }

    /**
     * a given class funtions as array keys
     * @param $class
     * @return array
     */
    protected static function parseClass($class)
    {
        $rows = array();
        foreach ($class['functions'] as $key => $value) {
            $row = array();
            $description = array();
            $parameters = array();
            $return = 'void';
            $row[] = $key;
            $value['doc'] = trim(str_replace(array("\r", "*", "/", '@', '|'), array('', '', '', '', '\|'), $value['doc']));
            $value['doc'] = explode("\n", $value['doc']);
            foreach ($value['doc'] as $line) {
                $line = trim(preg_replace('~[.[:cntrl:]]~', '', $line));
                if (preg_match('/^param/i', $line)) {
                    $line = str_replace('param ', '', $line);
                    $parameters[] = $line;
                } else if (preg_match('/^return/i', $line)) {
                    $line = str_replace('\|', '<br>', $line);
                    $line = str_replace('return ', '', $line);
                    $return = $line;
                } else {
                    $description[] = substr($line,0,50);
                }
            }
            $row[] = implode(' <br> ', $description);
            $row[] = implode(' <br> ', $value['modifiers']);
            $row[] = implode(' <br> ', $parameters);
            $row[] = $return;
            $rows[] = $row;
        }
        return $rows;
    }


    /**
     * a given class tags as markdown formated paragraph
     * @param $class
     * @return string
     */
    protected static function parseExtended($class)
    {
        $name = $class['name'];
        $autor = '';
        $description = NULL;
        $coping = '(c) '.date('Y');
        $version = '';
        $class['doc'] = trim(str_replace(array("\r", "*", "/", '|'), array('', '', '', '', '\|'), $class['doc']));
        $class['doc'] = explode("\n", $class['doc']);
        foreach ($class['doc'] as $line) {
            if (is_null($description)){
                $description = substr($line,0,55);
            }
            if (strpos($line, 'uthor') !== FALSE) {
                $line = str_replace('@author ', '', $line);
                $autor .= '* '.$line. PHP_EOL;
            }
            $line = str_replace('@', '', $line);
            if (strpos($line, 'ersion') !== FALSE) {
                $version = trim(preg_replace('~[[:cntrl:]]~', '', $line));
                $version = $version;
            }else {
                $line = trim(preg_replace('~[.[:cntrl:]]~', '', $line));
                if (preg_match('/^param/i', $line)) {
                    $line = str_replace('param ', '', $line);
                    $parameters = $line;
                } else if (preg_match('/^copyright/i', $line)) {
                    $line = str_replace('\|', '<br>', $line);
                    $line = str_replace('copyright ', '', $line);
                    $coping = $line;
                }
			}
        }
        $info = '**'.$name.', '.$version.'**: '.$description. PHP_EOL. PHP_EOL;
        $info .= $coping. PHP_EOL. PHP_EOL;
        $info .= $autor;
        return $info. PHP_EOL;
    }
    

    /**
     * Get markdown class documentation
     * @param $file
     * @return string
     */
    public static function getMarkdown($file)
    {
        $result = '';
        $cp = new ClassParser();
        $cp->parse($file);
        $classes = $cp->getClasses();
        foreach ($classes as $k => $class) {
            $class['name'] = $k;
            $result .= "## " . $class['name'] . PHP_EOL . PHP_EOL;
            $rows = self::parseClass($class);
            $columns = ['Method', 'Description', 'Type', 'Parameters', 'Return'];
            $result .= self::parseExtended($class);
            $t = new TextTable($columns, $rows);
            $result .= $t->render();
            $result .= PHP_EOL . PHP_EOL;
        }
        return $result;
    }

    /**
     * Print Markdown class documentation
     * @param $file
     */
    public static function printMarkdown($file)
    {
        echo self::getMarkdown($file);
    }

    /**
     * Get PHP file as array class documentation
     * @param $file
     * @return array
     */
    public static function getArray($file)
    {
        $result = array();
        $cp = new ClassParser();
        $cp->parse($file);
        $classes = $cp->getClasses();
        foreach ($classes as $k => $class) {
            $class['name'] = $k;
            $result[] = self::parseClass($class);
        }
        return $result;
    }
}
